### PR TITLE
fix(eslint-plugin): [no-unnecessary-condition] false positive when array predicate returns unknown

### PR DESCRIPTION
**Pull request by ArnaudBarre**
Fix: Mark unknown and any as possibly falsy.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-unnecessary-condition.ts [hunk 1]
```diff
@@ -33,6 +33,8 @@ const isTruthyLiteral = (type: ts.Type): boolean =>
   isBooleanLiteralType(type, true) || (isLiteralType(type) && !!type.value);
 
 const isPossiblyFalsy = (type: ts.Type): boolean =>
+  isTypeUnknownType(type) ||
+  isTypeAnyType(type) ||
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/no-unnecessary-condition.ts:37]: Making this change impacts the entirety of the rule, not just specifically the array predicate checks.

This change needs to be made such that it just impacts the required checks.

**Review by bradzacher — CHANGES_REQUESTED** [review]

File: packages/eslint-plugin/src/rules/no-unnecessary-condition.ts (hunk 1 above)
Comment (ArnaudBarre) [reply packages/eslint-plugin/src/rules/no-unnecessary-condition.ts:37]: This function is only used for that case and [in checkNode](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L254), 10 lines after an early return with [a similar condition](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L241-L242).

Also it seams always valid that any and unknown are possibly falsy (more complicated for type parameter)

If it's a performance consideration I can try to merge both conditions

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/no-unnecessary-condition.ts:37]: no point having one path check X and then rechecking X unnecessarily when the other code path just checks X once.

please standardise the two, or if that's not possible - localise this change closer to where it needs to be.

**Review by bradzacher — APPROVED** [review]
LGTM - thanks!

**Pull request merged**